File: uva/10054/Main.java
/*
 *  Copyright (c) 2014 Liang Li <ll@lianglee.org>. All rights reserved.
 *
 *  This program is a free software and released under the BSD license.
 *  https://github.com/lang010/acit
 *
 *  Solutions for UVa Problem 10054
 *  UVa link: http://uva.onlinejudge.org/external/100/10054.html
 *
 *  @Authur Liang Li <ll@lianglee.org>
 *  @Date Nov 16 16:06:34 2014
 */

import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.InputStreamReader;
import java.io.BufferedReader;

public class Main {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in, "ISO-8859-1"));
    PrintWriter pw = new PrintWriter(new OutputStreamWriter(System.out, "ISO-8859-1"));
    public Main() throws Exception {
    }

    public static void main(String[] args) throws Exception {
	Main m = new Main();
	m.run();
	m.release();
    }
    void release() throws Exception {
	br.close();
	pw.close();
    }

    int n;
    int[][] map = new int[100][100];
    int[] visited = new int[100];
    int[] degree = new int[100];

    int[] link = new int[1010];
    int nl;

    void init() {
        for (int i = 0; i < 100; i++) {
            for (int j = 0; j < 100; j++) {
                map[i][j] = 0;
            }
            degree[i] = 0;
            visited[i] = 0;
        }
    }

    int dfs(int id) {
        int r = 0;
        //pw.println("T: " + id);
        for (int i = 1; i <= 50; i++ ) {
            if (map[id][i] > 0) {
                map[id][i]--;
                map[i][id]--;
                //pw.println("TT: " + i);
                r += dfs(i) + 1;
            }
        }

        link[nl++] = id;
        return r;
    }

    void run() throws Exception {
        int t = Integer.valueOf(br.readLine());
        int x=0, y=0;
        for (int i = 0; i < t; i++) {
            init();
            n = Integer.valueOf(br.readLine());
            for (int j = 0; j < n; j++) {
                String s = br.readLine();
                String[] sl = s.split(" ");
                x = Integer.valueOf(sl[0]);
                y = Integer.valueOf(sl[1]);
                map[x][y]++;
                map[y][x]++;
                degree[x]++;
                degree[y]++;
            }
            boolean ok = true;
            for (int j= 0; j <= 50; j++) {
                if (degree[j]%2==1) {
                    ok = false;
                    break;
                }
            }
            nl = 0;
            if (ok) {
                for (int j = 0; j <= 50; j++) {
                    if (degree[j]>0) {
                        dfs(j);
                        break;
                    }
                }
            }
            if (i>0) {
                System.out.println();
            }
            System.out.printf("Case #%d%n", i+1);
            if (nl == n+1) {
                for (int j = 0; j < nl-1; j++) {
                    System.out.printf("%d %d%n", link[j], link[j+1]);
                }
            } else {
                System.out.println("some beads may be lost");
            }
        }

    }

}

File: uva/10054/main.cc
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int G[100][100];
int N;
int degree[100];
int link[1100];
int ln;

void dfs(int id) {
    for(int i = 1; i <= 50; i++) {
        if (G[id][i] > 0) {
            G[id][i]--;
            G[i][id]--;
            dfs(i);
        }
    }
    link[ln++] = id;
}

int main() {
    int T;
    int x,y,i;
    scanf("%d", &T);
    for( int k = 0; k < T; k++) {
        if (k>0)
            printf("\n");
        memset(G, 0, sizeof(G));
        memset(degree, 0, sizeof(degree));
        ln = 0;
        scanf("%d", &N);
        for(int i = 0; i < N; i++) {
            scanf("%d%d", &x, &y);
            degree[x]++;
            degree[y]++;
            G[x][y]++;
            G[y][x]++;
        }
        int ok = 1;
        for(i = 0; i <= 50; i++) {
            if (degree[i]%2 == 1) {
                ok = 0;
                break;
            }
        }
        printf("Case #%d\n", k+1);
        if (ok) {
            dfs(x);
            //printf("TT: %d\n", ln);
            if (ln != N+1)
                ok = 0;
            else {
                for(i = 0; i < N; i++)
                    printf("%d %d\n", link[i], link[i+1]);
            }
        }
        if (ok == 0) {
            printf("some beads may be lost\n");
        }
    }

    return 0;
}
